### PR TITLE
Create PostgreSQL feature

### PR DESCRIPTION
**Pull request by benmccann**
This refactoring sets the ground for allowing users to select other database types and follows the recommendation from https://github.com/micronaut-projects/micronaut-starter/issues/184#issuecomment-621390513

### CODE REVIEW

**Comment by benmccann** [comment]
It seems like the failing test is an unrelated flake. It's for Alexa and looks to be possibly related to a network failure. I wonder if there's anyway to turn off printing the name of every test that passes? It was really hard to find the failure in the logs because there was so much output

>io.micronaut.starter.generator.CreateAlexaSpec create-function-app with features [aws-alexa] java and maven and test framework: junit FAILED (2m)
  Condition not satisfied after 120.00 seconds and 118 attempts
      at spock.util.concurrent.PollingConditions.within(PollingConditions.java:170)
      at spock.util.concurrent.PollingConditions.eventually(PollingConditions.java:140)
      at io.micronaut.starter.generator.CommandSpec.testOutputContains(CommandSpec.groovy:53)
      at io.micronaut.starter.generator.CreateAlexaSpec.create-function-app with features #features #lang and #build and test framework: #testFramework(CreateAlexaSpec.groovy:55)
  Caused by: Condition not satisfied:
  output.toString().contains(value)
  |      |          |        |
  |      |          false    BUILD SUCCESS
  |      [INFO] Scanning for projects...
  |      [INFO] 
  |      [INFO] -----------------------< example.micronaut:foo >------------------------
  |      [INFO] Building foo 0.1
  |      [INFO] --------------------------------[ jar ]---------------------------------
  |      Downloading from jcenter.bintray.com: https://jcenter.bintray.com/io/micronaut/aws/micronaut-function-aws-alexa/2.0.0.M2/micronaut-function-aws-alexa-2.0.0.M2.pom
  |      Progress (1): 0.8/3.3 kB
  |      Progress (1): 2.2/3.3 kB
  |      Progress (1): 3.3 kB

File: starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java [hunk 1]
```diff
@@ -70,12 +70,8 @@ public void processSelectedFeatures(FeatureContext featureContext) {
     @Override
     public void apply(GeneratorContext generatorContext) {
         Map<String, Object> config = generatorContext.getConfiguration();
-        config.put("dataSource.url", "jdbc:h2:mem:devDb;MVCC=TRUE;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE");
         config.put("dataSource.pooled", true);
         config.put("dataSource.jmxExport", true);
-        config.put("dataSource.driverClassName", "org.h2.Driver");
-        config.put("dataSource.username", "sa");
-        config.put("dataSource.password", "");
```
Comment (jameskleeh) [inline starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:78]: This will definitely cause GORM applications to not function. We can't assume because the H2 feature is added that it is using `micronaut-jdbc` and expecting `datasources.default`. I think you'll need to leave the configuration up to the individual feature, however perhaps add a method or methods on the driver feature to make that easier.

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy [hunk 2]
```diff
@@ -71,7 +71,7 @@ class HibernateGormSpec extends BeanContextSpec {
         GeneratorContext ctx = buildGeneratorContext(['hibernate-gorm'])
 
         then:
-        ctx.configuration.containsKey("dataSource.url")
+        ctx.configuration.containsKey("datasources.default.url")
```
Comment (jameskleeh) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy:74]: This is incorrect

File: starter-core/src/main/java/io/micronaut/starter/feature/build/maven/templates/pom.rocker.raw [hunk 3]
```diff
@@ -434,6 +434,9 @@ List<Property> properties
 @if (features.contains("h2")) {
 @dependency.template("com.h2database", "h2", "runtime", null)
 }
+@if (features.contains("h2")) {
```
Comment (jameskleeh) [inline starter-core/src/main/java/io/micronaut/starter/feature/build/maven/templates/pom.rocker.raw:437]: This should be `.contains(postgressql)`

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy (hunk 2 above)
Comment (benmccann) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy]: fixed

File: starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java (hunk 1 above)
Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:78]: Fixed

Comment (jameskleeh) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:78]: This config is still being removed, so this will still cause GORM applications to not function

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy [hunk 4]
```diff
@@ -71,7 +71,7 @@ class HibernateGormSpec extends BeanContextSpec {
         GeneratorContext ctx = buildGeneratorContext(['hibernate-gorm'])
 
         then:
-        ctx.configuration.containsKey("dataSource.url")
+        ctx.configuration.containsKey("datasource.url")
```
Comment (jameskleeh) [inline starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy:74]: This is incorrect

File: starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java [hunk 5]
```diff
@@ -186,10 +186,13 @@ public void applyFeatures() {
     }
 
     public boolean isFeaturePresent(Class<? extends Feature> feature) {
-        Objects.requireNonNull(feature, "The feature class cannot be null");
```
Comment (jameskleeh) [inline starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java:189]: Why was this removed?

File: starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java (hunk 1 above)
Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java]: fixed

File: starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy (hunk 4 above)
Comment (benmccann) [reply starter-core/src/test/groovy/io/micronaut/starter/feature/database/HibernateGormSpec.groovy]: oops. fixed

File: starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java (hunk 5 above)
Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/application/generator/GeneratorContext.java:189]: `features.isFeaturePresent` has the same check, so it was redundant

File: starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java [hunk 6]
```diff
@@ -70,12 +70,16 @@ public void processSelectedFeatures(FeatureContext featureContext) {
     @Override
     public void apply(GeneratorContext generatorContext) {
         Map<String, Object> config = generatorContext.getConfiguration();
-        config.put("dataSource.url", "jdbc:h2:mem:devDb;MVCC=TRUE;LOCK_TIMEOUT=10000;DB_CLOSE_ON_EXIT=FALSE");
+        DatabaseDriverFeature dbFeature = generatorContext.getFeature(DatabaseDriverFeature.class);
+        if (dbFeature == null) {
+            dbFeature = defaultDbFeature;
+        }
```
Comment (jameskleeh) [inline starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:76]: The `getFeature` method should return an optional, with a `getRequiredFeature` that returns the feature and throws an exception if it doesn't exist. In this scenario we should expect that the feature exists.

Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:76]: Yeah, I had something close to that at first. The difficulty I ran into was that `FeatureSpec` calls `applyFeatures` on every feature without first calling `processSelectedFeatures`, so the database features aren't present and it would result in an exception.

Do you have any thoughts on how to handle that? Should I try to add a database feature always in `FeatureSpec`? That might not make sense for some features, but it might be messy to try to detect which features are the ones that require it and then add it. Or should I just leave the `null` handling here?

Comment (jameskleeh) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:76]: @benmccann The line in the test should probably be `getFeatures([feature.name]).getFeatures()` then

Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:76]: Sorry, I'm not following. Are you suggesting I go ahead and add `H2` to the `generatorContext` of for every feature in the test whether or not it's DB related?

Comment (benmccann) [reply starter-core/src/main/java/io/micronaut/starter/feature/database/HibernateGorm.java:76]: @jameskleeh thanks for the review! I'd love to close this out. Any advice on moving forward?
